lib/guardian/post_guardian.rb

```diff
@@ -102,14 +102,17 @@ def can_delete_all_posts?(user)
       user.post_count <= SiteSetting.delete_all_posts_max.to_i))
   end
 
-  def can_create_post?(parent)
-    return false if !SiteSetting.enable_system_message_replies? && parent.try(:subtype) == "system_message"
+  def can_create_post?(topic)
+    return can_create_post_in_topic?(topic) if !topic
 
-    (!SpamRule::AutoSilence.prevent_posting?(@user) || (!!parent.try(:private_message?) && parent.allowed_users.include?(@user))) && (
-      !parent ||
-      !parent.category ||
-      Category.post_create_allowed(self).where(id: parent.category.id).count == 1
-    )
+    key = topic_memoize_key(topic)
+    @can_create_post ||= {}
+
+    if @can_create_post.has_key?(key)
+      return @can_create_post[key]
+    end
+
+    @can_create_post[key] = can_create_post_in_topic?(topic)
```

Comment:
Maybe 'too magic', but `Hash#fetch` could simplify this a little:
```suggestion
    @can_create_post.fetch(key) do
      @can_create_post[key] = can_create_post_in_topic?(topic)
    end
```